File: Algorithms/pathSumII.py
class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    def pathSum(self, root, sum):
        self.result = []
        self.path = []
        self.pathSumHelper(root, sum)
        return self.result
    def pathSumHelper(self, root, sum):
        if not root: return
        self.path.append(root.val)
        if not root.left and not root.right and root.val == sum:
            self.result.append(self.path[:])
            self.path.pop()
            return
        self.pathSumHelper(root.left, sum - root.val)
        self.pathSumHelper(root.right, sum - root.val)
        self.path.pop()

File: Algorithms/pathSumII.java
import java.util.*;
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> pathSum(TreeNode root, int sum) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        ArrayList<Integer> path = new ArrayList<Integer>();
        path_sum(root, sum, path, result);
        return result;
    }
    void path_sum(TreeNode root, int sum, ArrayList<Integer> path, ArrayList<ArrayList<Integer>> result) {
        if (root == null) {
            return;
        }
        path.add(root.val);
        if (root.left == null && root.right == null && root.val == sum) {
            result.add((ArrayList<Integer>)path.clone());
            path.remove(path.size()-1);
            return;
        }
        path_sum(root.left, sum-root.val, path, result);
        path_sum(root.right, sum-root.val, path, result);
        path.remove(path.size()-1);
    }
}
